src/main/kotlin/treeLib/bintrees/RBTree.kt

```diff
@@ -0,0 +1,204 @@
+        val sonOfRemovedNode = when {
+            sonRight != null && sonLeft != null -> {
+                treeBranch.addFirst(curNode)
+                val copyTo = curNode    //Will copy from curNode
+                curNode = sonLeft
+                while (curNode.right != null) {
+                    treeBranch.addFirst(curNode)
+                    curNode = curNode.right ?: throw NullPointerException("Removing node cannot be null.")
+                }
+                parent = treeBranch.first()
+                val temp = copyTo.value
+                copyTo.key = curNode.key
+                copyTo.value = curNode.value
+                curNode.value = temp  //to save deleted node's value
+                curNode.left
+            }
+            sonRight != null -> sonRight
+            else /*sonLeft != null*/ -> sonLeft
+        }
+        amountOfNodes--
+        when (curNode) {   //here we are "deleting" the node and replace it with his son
+            parent?.right -> parent.right = sonOfRemovedNode
+            parent?.left -> parent.left = sonOfRemovedNode
+            else -> root = sonOfRemovedNode   //only root doesn't have parent
+        }
+        treeBranch.addFirst(sonOfRemovedNode)
+        //curNode.isRed -> no balancing needed
+        if (!curNode.isRed) balancedRemove(treeBranch)
+        return curNode.value
+    }
+
+    /**
+     * Used in remove() method to balance tree :(
+     */
+    private fun balancedRemove(treeBranch: ArrayDeque<RBNode<K, V>?>) {
+        var son = treeBranch.removeFirstOrNull()
+        var parent = treeBranch.removeFirstOrNull()
+        var grandparent = treeBranch.removeFirstOrNull()
+        //son == son of removed node
+        if (son?.isRed == true) {
+            son.isRed = false
+            return
+        }
+        while (parent != null) {
+            var brother = if (son == parent.right) parent.left else parent.right
+            //we want for brother to be black (´。＿。｀)
+            if (brother?.isRed == true) {
+                brother.isRed = false
+                parent.isRed = true
+                if (son == parent.right) {
+                    rotateRight(parent, grandparent)
+                    grandparent = brother
+                    brother = parent.left
+                } else /*son == parent.left*/ {
+                    rotateLeft(parent, grandparent)
+                    grandparent = brother
+                    brother = parent.right
+                }
+            }
+            brother ?: throw IllegalStateException(
+                "Balance of red-black tree violated:" +
+                    " parent node have subtrees with black height 2 and black height 1"
+            )
+            //depending on nephew's colors, doing magic for balancing purposes
+            var nephewRight = brother.right
+            var nephewLeft = brother.left
+            if ((nephewRight == null || !nephewRight.isRed) && (nephewLeft == null || !nephewLeft.isRed)) {
```

Comment:
Предлагаю выделить
```kotlin
(nephewRight == null || !nephewRight.isRed)
```
В отдельный метод, потому что условие на black node применяется в 4 местах и введение нового метода увеличит читаемость и спасет от потенциальных багов

```suggestion
    private fun isBlack(node: RBNode<K, V>?): Boolean {
        return node == null || !node.isRed
    }
    
    if (isBlack(nephewRight) && isBlack(nephewLeft)) {
```